File: backend/MachineryCRM.Application/Services/CustomerService.cs
using MachineryCRM.Application.DTOs;
using MachineryCRM.Application.Interfaces;
using MachineryCRM.Domain.Entities;
using MachineryCRM.Domain.Interfaces;

namespace MachineryCRM.Application.Services;

public class CustomerService : ICustomerService
{
    private readonly ICustomerRepository _customerRepository;
    private readonly IUnitOfWork _unitOfWork;

    public CustomerService(ICustomerRepository customerRepository, IUnitOfWork unitOfWork)
    {
        _customerRepository = customerRepository;
        _unitOfWork = unitOfWork;
    }

    public async Task<CustomerDto> CreateAsync(CreateCustomerDto dto)
    {
        var customer = new Customer(dto.Name);
        
        await _customerRepository.AddAsync(customer);
        await _unitOfWork.CommitAsync();

        return new CustomerDto { Id = customer.Id, Name = customer.Name };
    }

public async Task<IEnumerable<CustomerDto>> GetAllAsync()
{
    var customers = await _customerRepository.GetAllWithDetailsAsync();
    
    return customers.Select(c => new CustomerDto 
    { 
        Id = c.Id, 
        Name = c.Name,
        // O mapeamento das listas é obrigatório para que o frontend consiga calcular o ".length"
        Sites = c.Sites.Select(s => new SiteDto { Id = s.Id, Name = s.Name, Country = s.Country, State = s.State, City = s.City }).ToList(),
        FiscalEntities = c.FiscalEntities.Select(f => new FiscalEntityDto { Id = f.Id, Name = f.Name, Country = f.Country, State = f.State, City = f.City }).ToList(),
        Contacts = c.Contacts.Select(ct => new ContactDto { Id = ct.Id, Description = ct.Description }).ToList()
    });
}

public async Task<CustomerDto?> GetByIdWithDetailsAsync(Guid id)
{
    var customer = await _customerRepository.GetCustomerWithDetailsAsync(id);
    if (customer == null) return null;

    var customerDto = new CustomerDto
    {
        Id = customer.Id,
        Name = customer.Name,
        Sites = customer.Sites.Select(s => new SiteDto 
        { 
            Id = s.Id, Name = s.Name, Country = s.Country, State = s.State, City = s.City, Observations = s.Observations 
        }).ToList(),
        FiscalEntities = customer.FiscalEntities.Select(f => new FiscalEntityDto 
        { 
            Id = f.Id, Name = f.Name, SapPn = f.SapPn, Cnpj = f.Cnpj, Cpf = f.Cpf, Country = f.Country, State = f.State, City = f.City 
        }).ToList(),
        Contacts = customer.Contacts.Select(c => new ContactDto 
        { 
            Id = c.Id, SiteId = c.SiteId, FiscalEntityId = c.FiscalEntityId, Description = c.Description, Phone = c.Phone, Email = c.Email 
        }).ToList()
    };

    return customerDto;
}

public async Task<SiteDto> AddSiteAsync(Guid customerId, CreateSiteDto dto)
{
    // Busca leve: carrega apenas o cliente, sem as listas pesadas
    var customer = await _customerRepository.GetByIdAsync(customerId);
    if (customer == null) throw new KeyNotFoundException("Cliente não encontrado.");

    var site = new Site(customerId, dto.Name, dto.Country, dto.State, dto.City);
    
    // Inserção direta sem modificar a entidade Customer
    _customerRepository.AddSite(site);
    await _unitOfWork.CommitAsync();

    return new SiteDto 
    { 
        Id = site.Id, Name = site.Name, Country = site.Country, State = site.State, City = site.City 
    };
}

public async Task<FiscalEntityDto> AddFiscalEntityAsync(Guid customerId, CreateFiscalEntityDto dto)
{
    var customer = await _customerRepository.GetByIdAsync(customerId);
    if (customer == null) throw new KeyNotFoundException("Cliente não encontrado.");

    var fiscal = new FiscalEntity(customerId, dto.Name, dto.Country, dto.State, dto.City);
    
    // Inserção direta sem modificar a entidade Customer
    _customerRepository.AddFiscalEntity(fiscal);
    await _unitOfWork.CommitAsync();

    return new FiscalEntityDto 
    { 
        Id = fiscal.Id, Name = fiscal.Name, Country = fiscal.Country, State = fiscal.State, City = fiscal.City 
    };
}

public async Task<ContactDto> AddContactAsync(Guid customerId, CreateContactDto dto)
{
    var customer = await _customerRepository.GetByIdAsync(customerId);
    if (customer == null) throw new KeyNotFoundException("Cliente não encontrado.");

    var contact = new Contact(customerId, dto.Description, dto.Phone, dto.Email);
    
    // Se a entidade utilizar métodos para atribuir chaves estrangeiras opcionais:
    if (dto.SiteId.HasValue) contact.ChangeSite(dto.SiteId.Value);
    if (dto.FiscalEntityId.HasValue) contact.ChangeFiscalEntity(dto.FiscalEntityId.Value);
    
    _customerRepository.AddContact(contact);
    await _unitOfWork.CommitAsync();

    return new ContactDto 
    { 
        Id = contact.Id, 
        SiteId = contact.SiteId, 
        FiscalEntityId = contact.FiscalEntityId, 
        Description = contact.Description, 
        Phone = contact.Phone, 
        Email = contact.Email 
    };
}

// --- CUSTOMER ---
public async Task UpdateCustomerAsync(Guid id, UpdateCustomerDto dto)
{
    var customer = await _customerRepository.GetByIdAsync(id);
    if (customer == null) throw new KeyNotFoundException("Cliente não encontrado.");
    customer.UpdateName(dto.Name);
    _customerRepository.Update(customer);
    await _unitOfWork.CommitAsync();
}

public async Task DeleteCustomerAsync(Guid id)
{
    var customer = await _customerRepository.GetByIdAsync(id);
    if (customer == null) throw new KeyNotFoundException("Cliente não encontrado.");
    _customerRepository.Remove(customer); // A exclusão em cascata deve cuidar dos filhos se configurada no EF
    await _unitOfWork.CommitAsync();
}

// --- SITE ---
public async Task UpdateSiteAsync(Guid siteId, UpdateSiteDto dto)
{
    var site = await _customerRepository.GetSiteByIdAsync(siteId);
    if (site == null) throw new KeyNotFoundException("Local não encontrado.");
    site.UpdateDetails(dto.Name, dto.Country, dto.State, dto.City);
    await _unitOfWork.CommitAsync(); // O EF rastreia a mudança da entidade carregada
}

public async Task DeleteSiteAsync(Guid siteId)
{
    var site = await _customerRepository.GetSiteByIdAsync(siteId);
    if (site == null) throw new KeyNotFoundException("Local não encontrado.");
    _customerRepository.RemoveSite(site);
    await _unitOfWork.CommitAsync();
}

// --- FISCAL ENTITY ---
public async Task UpdateFiscalEntityAsync(Guid fiscalId, UpdateFiscalEntityDto dto)
{
    var fiscal = await _customerRepository.GetFiscalEntityByIdAsync(fiscalId);
    if (fiscal == null) throw new KeyNotFoundException("Ente fiscal não encontrado.");
    fiscal.UpdateDetails(dto.Name, dto.SapPn, dto.Cnpj, dto.Cpf, dto.Country, dto.State, dto.City);
    await _unitOfWork.CommitAsync();
}

public async Task DeleteFiscalEntityAsync(Guid fiscalId)
{
    var fiscal = await _customerRepository.GetFiscalEntityByIdAsync(fiscalId);
    if (fiscal == null) throw new KeyNotFoundException("Ente fiscal não encontrado.");
    _customerRepository.RemoveFiscalEntity(fiscal);
    await _unitOfWork.CommitAsync();
}

// --- CONTACT ---
public async Task UpdateContactAsync(Guid contactId, UpdateContactDto dto)
{
    var contact = await _customerRepository.GetContactByIdAsync(contactId);
    if (contact == null) throw new KeyNotFoundException("Contato não encontrado.");
    
    contact.UpdateDetails(dto.Description, dto.Phone, dto.Email, null);
    
    // Lógica para permitir arrastar/mover contatos entre cards no frontend
    if (contact.SiteId != dto.SiteId) contact.ChangeSite(dto.SiteId);
    if (contact.FiscalEntityId != dto.FiscalEntityId) contact.ChangeFiscalEntity(dto.FiscalEntityId);

    await _unitOfWork.CommitAsync();
}

public async Task DeleteContactAsync(Guid contactId)
{
    var contact = await _customerRepository.GetContactByIdAsync(contactId);
    if (contact == null) throw new KeyNotFoundException("Contato não encontrado.");
    _customerRepository.RemoveContact(contact);
    await _unitOfWork.CommitAsync();
}
}

File: frontend/src/services/customerService.ts
import { api } from './api';

export interface SiteDto {
  id: string;
  name: string;
  country: string;
  state: string;
  city: string;
  observations?: string;
}

export interface FiscalEntityDto {
  id: string;
  sapPn?: string;
  name: string;
  cnpj?: string;
  cpf?: string;
  country: string;
  state: string;
  city: string;
}

export interface ContactDto {
  id: string;
  siteId?: string;
  fiscalEntityId?: string;
  description: string;
  phone?: string;
  email?: string;
}

export interface CustomerDto {
  id: string;
  name: string;
  sites: SiteDto[];
  fiscalEntities: FiscalEntityDto[];
  contacts: ContactDto[];
}

export interface CreateCustomerDto {
  name: string;
}

export interface CreateSiteDto {
  name: string;
  country: string;
  state: string;
  city: string;
  observations?: string;
}

export interface CreateFiscalEntityDto {
  sapPn?: string;
  name: string;
  cnpj?: string;
  cpf?: string;
  country: string;
  state: string;
  city: string;
}

export interface CreateContactDto {
  siteId?: string;
  fiscalEntityId?: string;
  description: string;
  phone?: string;
  email?: string;
}

export const GeoLocationType = {
  Office: 0,
  MachineLocation: 1,
  Waypoint: 2
} as const;

export type GeoLocationType = (typeof GeoLocationType)[keyof typeof GeoLocationType];

export interface GeoPointDto {
  id: string;
  description: string;
  latitude: number;
  longitude: number;
  locationType: GeoLocationType;
  order?: number;
}

export interface CreateGeoPointDto {
  description: string;
  latitude: number;
  longitude: number;
  locationType: GeoLocationType;
  order?: number;
}

export interface UpdateGeoPointDto {
  description: string;
  latitude: number;
  longitude: number;
}

export interface ReorderGeoPointDto {
  id: string;
  order: number;
}

// Atualize a interface SiteDto existente para incluir geoPoints:
export interface SiteDto {
  id: string;
  name: string;
  country: string;
  state: string;
  city: string;
  observations?: string;
  geoPoints?: GeoPointDto[];
}

export const getCustomers = async (): Promise<CustomerDto[]> => {
  const response = await api.get<CustomerDto[]>('/customers');
  return response.data;
};

export const getCustomerById = async (id: string): Promise<CustomerDto> => {
  const response = await api.get<CustomerDto>(`/customers/${id}`);
  return response.data;
};

export const createCustomer = async (data: CreateCustomerDto): Promise<CustomerDto> => {
  const response = await api.post<CustomerDto>('/customers', data);
  return response.data;
};

export const addSiteToCustomer = async (customerId: string, data: CreateSiteDto): Promise<SiteDto> => {
  const response = await api.post<SiteDto>(`/customers/${customerId}/sites`, data);
  return response.data;
};

export const addFiscalEntityToCustomer = async (customerId: string, data: CreateFiscalEntityDto): Promise<FiscalEntityDto> => {
  const response = await api.post<FiscalEntityDto>(`/customers/${customerId}/fiscal-entities`, data);
  return response.data;
};

export const addContactToCustomer = async (customerId: string, data: CreateContactDto): Promise<ContactDto> => {
  const response = await api.post<ContactDto>(`/customers/${customerId}/contacts`, data);
  return response.data;
};
export const addGeoPointToSite = async (siteId: string, data: CreateGeoPointDto): Promise<GeoPointDto> => {
  const response = await api.post<GeoPointDto>(`/customers/sites/${siteId}/geopoints`, data);
  return response.data;
};

export const updateGeoPoint = async (geoPointId: string, data: UpdateGeoPointDto): Promise<void> => {
  await api.put(`/customers/geopoints/${geoPointId}`, data);
};

export const deleteGeoPoint = async (geoPointId: string): Promise<void> => {
  await api.delete(`/customers/geopoints/${geoPointId}`);
};

export const reorderGeoPoints = async (siteId: string, data: ReorderGeoPointDto[]): Promise<void> => {
  await api.put(`/customers/sites/${siteId}/geopoints/reorder`, data);
};